### ISSUE TITLE
Breadcrumbs: clues written to the test output if a test does not pass

### ISSUE THREAD

**Issue by AGBrown** [nunit/nunit#2915]
#### Summary

It would be useful to be able to output a set of state records alongside a failing assertion without having to explicitly include these details in the failure message in the assert statement. For example: have an implicit "Breadcrumbs" collection scoped to a running Test which is written out alongside the failure details. This is much like using `WriteLine`, but it would only be displayed if the test fails, therefore keeping test outputs and reports clean, and avoiding distraction.

##### What are breadcrumbs?

In the Sentry exception tracking tool there is the [concept of "breadcrumbs"][br]. These are values that are constantly written into a context. If an exception is thrown then all breadcrumbs (hints about state that are a step beyond a simple stack trace) are logged alongside the exception. 

##### Use case - non-specific input arguments

When using [constrained non-determinism][cnd] (using https://github.com/AutoFixture/AutoFixture), input values are not specified, instead the set of values that fall in the domain of the sut are specified and the argument is generated from this set at runtime.

[cnd]: http://blog.ploeh.dk/2009/03/05/ConstrainedNon-Determinism/

When tests fail it is often useful to know the argument value used for the failing case, whether it is always or just intermittently failing. 

In this case the concept of breadcrumbs would be very useful. During the arrange step test authors could drop breadcrumbs corresponding to: values generated with `fixture.Create<T>()`; intermediate variables returned from other functions; and for final arguments used as inputs.

##### Use case - domain of function coverage testing

The constrained non-determinism approach is distinct from and complementary to something like https://fscheck.github.io/FsCheck/ which performs coverage testing across the domain of a sut; Breadcrumbs would be similarly useful with FsCheck.

##### Use case - integration tests

In Integration tests this would also be useful - what specific Id of a data entity was the failure on. Can we now go and look at the database, and that exact record(s) to see what went wrong?

##### Use case - Plugin packages

It would be easy to foresee an Autofixture.NUnit.Breadcrumbs package which automatically added breadcrumbs for each call to `fixture.Create<T>()`, for instance - so that test authors didn't even have to know it had happened until there was a test failure - and all the input arguments were displayed along with the test failure messages in the test window. This would presumably leverage the same Breadcrumbs collection in the test context as test authors could access directly.

### TL;DR

For each running test have a Breadcrumbs object in the context for the test that test authors and plugins can write to. If (and only if) the test does not Pass, write these Breadcrumbs out alongside the test failure message and stack trace. Do not write anything out for tests that Pass so as to keep the environment and reports clean and distraction-free.

[br]: https://docs.sentry.io/clients/csharp/#breadcrumbs

**Comment by jnm2** [comment]
Thanks for the nice writeup!

What do you think about the option to write breadcrumbs to the test result XML while only displaying it in the log output on failure? That way I can look back through the saved result XML for previous succeeding CI builds and compare those breadcrumbs, but I don't have to watch all the messages scrolling past in the console when things are succeeding.

**Comment by AGBrown** [comment]
>  I can look back through the saved result XML for previous succeeding CI builds and compare those breadcrumbs

Interesting. My immediate thoughts were: 

- can we do that (write breadcrumbs for all tests), but only if one fails - thus keeping the XML tidy if there are no issues
- iterated to: can we do that, but only for passing tests relevant to the failing tests
- iterated to: but there may be multiple runs, and one test fails in one run, so the relevant "pass" breadcrumbs are in the previous test runs for the same test
- iterated to: so it looks like we need them all the time in the XML output, for all tests, so could we instead give the user the option of having "pass" breadcrumbs included or not such that they could have a fail test, the breadcrumbs for that fail written, and then turn on pass breadcrumbs to help identify inputs that work vs those that don't.

I think that's where I end up: I like your idea. Ideally it would be implemented with a switch like a logging scale: Off, ForFailingTestsOnly, ForAllTestsAlways

**Comment by CharliePoole** [comment]
@jnm2 A point to clarify: you refer to "only displaying it in the log output" and @AGBrown followed up on that comment. What did you mean by "log output". Our only output that we specifically call "log" is the internal trace log, which is rarely looked at by users unless there is an nunit bug.

**Comment by jnm2** [comment]
Would it make sense to default to Always, and add Off, FailingTestsOnly, and IfAnyFail as needed? Or would you really be happiest having those options from the beginning?

@CharliePoole By "log output" I meant console output.

**Comment by AGBrown** [comment]
> If we had the breadcrumbs feature, what would the XML look like? Specifically, are we talking about a separate element or including in the output for the test itself? Implementation would be very different.

@CharliePoole - I defer to the experts. I don't see a need for structured output at this point - structured output being a separate element, potentially with some kind of child element collection representing the keys/values. That may change in the future if the output is to be consumed in some way, but that's not really what matters here. A concatenated `key=value` message is grep-able, so feels ok right now. I don't think it needs over-complicating.

**Comment by AGBrown** [comment]
>  you refer to "only displaying it in the log output" and @AGBrown followed up on that comment.

> By "log output" I meant console output.

By "log output" I meant "I'm aware there are many outputs and I don't truly understand any of them, so again I'll defer to the experts". Our current use case would be to see these breadcrumbs in the VS test outputs (native or resharper) and in the report output file if we're running on the console.

**Comment by AGBrown** [comment]
FYI: we're currently working around this using NLog, [NLogViewer](https://github.com/NLog/NLog/wiki/NLogViewer-target) target, and something like Log4View, with NLog's [MDLC](https://github.com/NLog/NLog/wiki/Mdlc-Layout-Renderer) etc. for the "breadcrumbs". It's not quite the same, but it does work, and the code under test also logs to nlog so we see both test and SUT logs in the same place. There is some extra code to write to make this happen in the tests, but it's not too painful.
